src/topology/noetherian_space.lean

```diff
@@ -0,0 +1,220 @@
+    refine embedding_subtype_coe.is_compact_iff_is_compact_image.mpr (H (coe '' t) _),
+    simp }
+end
+
+@[simp] lemma noetherian_univ_iff :
+  noetherian_space (set.univ : set α) ↔ noetherian_space α :=
+noetherian_space_iff_of_homeomorph (homeomorph.set.univ α)
+
+lemma noetherian_space.Union {ι : Type*} [fintype ι] (f : ι → set α)
+  [hf : ∀ i, noetherian_space (f i)] :
+  noetherian_space (⋃ i, f i) :=
+begin
+  simp_rw noetherian_space_set_iff at hf ⊢,
+  intros t ht,
+  rw [← set.inter_eq_left_iff_subset.mpr ht, set.inter_Union],
+  exact compact_Union (λ i, hf i _ (set.inter_subset_right _ _))
+end
+
+-- This is not an instance since it makes a loop with `t2_space_discrete`.
+lemma noetherian_space.discrete [noetherian_space α] [t2_space α] : discrete_topology α :=
+⟨eq_bot_iff.mpr (λ U _, is_closed_compl_iff.mp (noetherian_space.is_compact _).is_closed)⟩
+
+local attribute [instance] noetherian_space.discrete
+
+/-- Spaces that are both Noetherian and Hausdorff is finite. -/
+@[priority 100]
+noncomputable
+def noetherian_space.fintype [noetherian_space α] [t2_space α] : fintype α :=
+set.fintype_of_finite_univ (noetherian_space.is_compact set.univ).finite_of_discrete
+
+@[priority 100]
+instance noetherian_space.of_fintype [fintype α] : noetherian_space α :=
+by { classical, exact ⟨@@fintype.well_founded_of_trans_of_irrefl (subtype.fintype _) _ _ _⟩ }
+
+lemma noetherian_space.exists_finset_irreducible [noetherian_space α] (s : closeds α) :
+  ∃ S : finset (closeds α), (∀ k : S, is_irreducible (k : set α)) ∧ s = S.sup id :=
+begin
+  classical,
+  have := ((noetherian_space_tfae α).out 0 1).mp infer_instance,
+  apply well_founded.induction this s, clear s,
+  intros s H,
+  by_cases h₁ : is_preirreducible s.1,
+  cases h₂ : s.1.eq_empty_or_nonempty,
+  { use ∅, refine ⟨λ k, k.2.elim, _⟩, rw finset.sup_empty, ext1, exact h },
+  { use {s},
+    simp only [coe_coe, finset.sup_singleton, id.def, eq_self_iff_true, and_true],
+    rintro ⟨k, hk⟩,
+    cases finset.mem_singleton.mp hk,
+    exact ⟨h, h₁⟩ },
+  { rw is_preirreducible_iff_closed_union_closed at h₁,
+    push_neg at h₁,
+    obtain ⟨z₁, z₂, hz₁, hz₂, h, hz₁', hz₂'⟩ := h₁,
+    obtain ⟨S₁, hS₁, hS₁'⟩ := H (s ⊓ ⟨z₁, hz₁⟩) (inf_lt_left.2 hz₁'),
+    obtain ⟨S₂, hS₂, hS₂'⟩ := H (s ⊓ ⟨z₂, hz₂⟩) (inf_lt_left.2 hz₂'),
+    refine ⟨S₁ ∪ S₂, λ k, _, _⟩,
+    { cases finset.mem_union.mp k.2 with h' h', exacts [hS₁ ⟨k, h'⟩, hS₂ ⟨k, h'⟩] },
+    { rwa [finset.sup_union, ← hS₁', ← hS₂', ← inf_sup_left, left_eq_inf] } }
+end
+
+lemma noetherian_space.finite_irreducible_components [noetherian_space α] :
+  (set.range irreducible_component : set (set α)).finite :=
+begin
+  classical,
+  obtain ⟨S, hS₁, hS₂⟩ := noetherian_space.exists_finset_irreducible (⊤ : closeds α),
+  suffices : ∀ x : α, ∃ s : S, irreducible_component x = s,
+  { choose f hf,
+    rw [show irreducible_component = coe ∘ f, from funext hf, set.range_comp],
+    exact set.finite.image _ (set.finite.intro infer_instance) },
```

Comment:
```suggestion
    exact (set.finite.intro _).image _ },
```

Reply:
I used the dot notation but the `infer_instance` still needs to be provided manually.

Reply:
Ah because `set.finite.intro` doesn't take `fintype s` as an instance argument. We need an alternative constructor that does.